### PR TITLE
feat(load-balancer): Add new node-selector annotation

### PR DESCRIPTION
**Pull request by xoxys**
Fixes: https://github.com/hetznercloud/hcloud-cloud-controller-manager/issues/506, https://github.com/hetznercloud/hcloud-cloud-controller-manager/issues/373

### CODE REVIEW

File: hcloud/load_balancers.go [hunk 1]
```diff
@@ -44,6 +45,29 @@ func newLoadBalancers(lbOps LoadBalancerOps, ac hcops.HCloudActionClient, disabl
...
+	var (
+		err           error
+		selectedNodes []*corev1.Node
+	)
+
+	selector := labels.Everything()
+	if v, ok := annotation.LBNodeSelector.StringFromService(svc); ok {
+		selector, err = labels.Parse(v)
+		if err != nil {
+			return nil, err
```
Comment (apricote) [inline hcloud/load_balancers.go:58]: An error here is exposed to the user as a `Event`. IMO its hard to understand with the current message where the issue actually happened, and what the user should do to fix this.

Error message from event: `Error syncing load balancer: failed to ensure load balancer: hcloud/loadBalancers.EnsureLoadBalancer: found '==', expected: !, identifier, or 'end of string'`

I would prefer if we added something along the lines of "unable to parse node-selector annotation:" to the error:

```suggestion
			return nil, fmt.Errorf("unable to parse the node-selector annotation: %w", err)
```

File: hcloud/load_balancers.go [hunk 2]
```diff
@@ -97,13 +121,19 @@ func (l *loadBalancers) EnsureLoadBalancer(
...
+		lb            *hcloud.LoadBalancer
+		err           error
+		selectedNodes []*corev1.Node
 	)
 
-	nodeNames := make([]string, len(nodes))
-	for i, n := range nodes {
+	selectedNodes, err = matchNodeSelector(svc, nodes)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %v", op, err)
```
Comment (apricote) [inline hcloud/load_balancers.go:132]:
```suggestion
		return nil, fmt.Errorf("%s: %w", op, err)
```

If you wrap errors, consumers can actually unwrap & inspect the the error later on if thats what they want to do.

https://go.dev/blog/go1.13-errors

File: hcloud/load_balancers.go [hunk 3]
```diff
@@ -236,12 +266,18 @@ func (l *loadBalancers) UpdateLoadBalancer(
...
+		lb            *hcloud.LoadBalancer
+		err           error
+		selectedNodes []*corev1.Node
 	)
 
-	nodeNames := make([]string, len(nodes))
-	for i, n := range nodes {
+	selectedNodes, err = matchNodeSelector(svc, nodes)
+	if err != nil {
+		return fmt.Errorf("%s: %v", op, err)
```
Comment (apricote) [inline hcloud/load_balancers.go:276]:
```suggestion
		return fmt.Errorf("%s: %w", op, err)
```

File: internal/annotation/load_balancer.go [hunk 4]
```diff
@@ -98,6 +98,10 @@ const (
 	// Mutually exclusive with LBLocation.
 	LBNetworkZone Name = "load-balancer.hetzner.cloud/network-zone"
 
+	// LBNodeSelector specifies a Kubernetes label selector that matches the
+	// string based format. This is used to select the nodes added to the Load Balancer.
```
Comment (apricote) [inline internal/annotation/load_balancer.go:102]:
```suggestion
	// LBNodeSelector can be set to restrict which Nodes are added as targets to the
	// Load Balancer. It accepts a Kubernetes label selector string, using either the
	// set-based or equality-based formats. 
	//
	// If the selector can not be parsed, the targets in the Load Balancer are not
	// updated and an Event is created with the error message.
	//
	// Format: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
```

**Review by apricote — CHANGES_REQUESTED** [review]
Thank you very much @xoxys! This works very well in my tests, I only have some UX suggestions left :)

**Comment by apricote** [comment]
Dont worry about the failing e2e tests. They are currently broken for external contributors. I ran them locally and everything passed :)

File: hcloud/load_balancers.go (hunk 1 above)
Comment (xoxys) [reply hcloud/load_balancers.go]: Ah yeah, that makes sense :+1: just copied the snippet and haven't thought enough about it :see_no_evil:

**Review by apricote — APPROVED** [review]

**Pull request merged**